bytestring/common/src/ByteString.kt

```diff
@@ -0,0 +1,453 @@
+ *
+ * Behavior of this method is compatible with [CharSequence.startsWith].
+ *
+ * @param byteArray the prefix to check for.
+ */
+public fun ByteString.startsWith(byteArray: ByteArray): Boolean = when {
+    byteArray.size > size -> false
+    else -> rangeEquals(0, byteArray)
+}
+
+/**
+ * Returns true if this byte string starts with the prefix specified by the [byteString].
+ *
+ * Behavior of this method is compatible with [CharSequence.startsWith].
+ *
+ * @param byteString the prefix to check for.
+ */
+public fun ByteString.startsWith(byteString: ByteString): Boolean = when {
+    byteString.size > size -> false
+    byteString.size == size -> equals(byteString)
+    else -> rangeEquals(0, byteString)
+}
+
+/**
+ * Returns true if this byte string ends with the suffix specified by the [byteArray].
+ *
+ * Behavior of this method is compatible with [CharSequence.endsWith].
+ *
+ * @param byteArray the suffix to check for.
+ */
+public fun ByteString.endsWith(byteArray: ByteArray): Boolean = when {
+    byteArray.size > size -> false
+    else -> rangeEquals(size - byteArray.size, byteArray)
+}
+
+/**
+ * Returns true if this byte string ends with the suffix specified by the [byteString].
+ *
+ * Behavior of this method is compatible with [CharSequence.endsWith].
+ *
+ * @param byteString the suffix to check for.
+ */
+public fun ByteString.endsWith(byteString: ByteString): Boolean = when {
+    byteString.size > size -> false
+    byteString.size == size -> equals(byteString)
+    else -> rangeEquals(size - byteString.size, byteString)
+}
+
+private fun ByteString.rangeEquals(
+    offset: Int, other: ByteString, otherOffset: Int = 0,
+    byteCount: Int = other.size - otherOffset
+): Boolean {
+    for (i in 0 until byteCount) {
+        if (this[offset + i] != other[otherOffset + i]) {
+            return false
+        }
+    }
+    return true
+}
+
+private fun ByteString.rangeEquals(
+    offset: Int, other: ByteArray, otherOffset: Int = 0,
+    byteCount: Int = other.size - otherOffset
+): Boolean {
+    for (i in 0 until byteCount) {
+        if (this[offset + i] != other[otherOffset + i]) {
+            return false
+        }
+    }
+    return true
+}
+
+/**
+ * Returns `true` if this byte string is empty.
+ */
+public fun ByteString.isEmpty(): Boolean = size == 0
+
+/**
+ * Returns `true` if this byte string is not empty.
+ */
+public fun ByteString.isNotEmpty(): Boolean = !isEmpty()
+
+/**
+ * Decodes content of a byte string into a string using UTF-8 encoding.
+ */
+public fun ByteString.toUtf8String(): String {
+    return getBackingArrayReference().decodeToString()
+}
+
+/**
+ * Encodes a string into a byte sequence using UTF8-encoding and wraps it into a byte string.
+ *
+ * @param string the string to be encoded.
+ */
+public fun ByteString.Companion.fromUtf8String(string: String): ByteString {
+    return wrap(string.encodeToByteArray())
```

Comment:
`encodeToByteArray` behaves differently on JVM and Native when encoding invalid code points.

It may be fine, but primitives from the code module behave similarly on all targets (and the behavior is the same as `encodeToByteArray`'s on JVM).

So we have to either hoist UTF-8 support from the code module into a separate module and reuse it here or change the behavior in the core module to match stdlib's behavior.

Either way, I'd rather postpone a fix and discuss it separately.

Reply:
Note: K/JS behaviour of `encodeToByteArray` is the same as in K/N (not sure about K/WASM). In that case only K/JVM behaviour is differs from other platforms.
And now in `kotlinx-io` we adopting K/JVM behaviour.
While it could be fine, it's a little inconsistent